### DISCUSSION TITLE
Just cant get this working , can any one give me some ideas

### DISCUSSION THREAD

**Discussion by TonyBrock42** [DCC-EX/EX-T3-WiFi#108] [Q&A] [answered]
Hi I have built this using the exact parts in the doc (no accelerometer).  Installed VSCode on a new clean Windows 10 machine, opened the unzipped T£ Wifi Master zip.  Compiled ok.  Added .json and icons to a 4gb SD and inserted into screen.  Did Upload File System Image (completed ok) then did Upload and Monitor and i get 
Wrote 1333392 bytes (802457 compressed) at 0x00010000 in 19.3 seconds (effective 553.1 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...

Please build project in debug configuration to get more details about an exception.
See https://docs.platformio.org/page/projectconf/build_configurations.html

--- Terminal on COM4 | 115200 8-N-1
--- Available filters and text transformations: colorize, debug, default, direct, esp32_exception_decoder, hexlify, log2file, nocontrol, printable, send_on_enter, time
--- More details at https://bit.ly/pio-monitor-filters
--- Quit: Ctrl+C | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H
ets Jul 29 2019 [12:21:46](https://www.youtube.com/watch?v=Z5gXmK8kCI4&t=44506s)

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13104
load:0x40080400,len:3036
entry 0x400805e4
[   677][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
[  1178][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
[  1779][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
[  2279][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
[  2880][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
[  3380][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
[  3981][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
[  4481][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
[  4481][E][vfs_api.cpp:24] open(): File system is not mounted
[  4482][E][vfs_api.cpp:24] open(): File system is not mounted
[  4487][E][esp32-hal-cpu.c:110] addApbChangeCallback(): duplicate func=0x40104738 arg=0x3ffbde68
[  6978][E][Wire.cpp:513] requestFrom(): i2cRead returned Error 263
[  7983][E][Wire.cpp:513] requestFrom(): i2cRead returned Error 263
[  8987][E][Wire.cpp:513] requestFrom(): i2cRead returned Error 263
[  9991][E][Wire.cpp:513] requestFrom(): i2cRead returned Error 263

I am running off USB C no battery connected.  Can you help at all please.

**Comment by alex-code** [comment]
Looks like it's trying to load.
Was the SD card formatted with FAT32?

**Reply by TonyBrock42** [reply]
Hi thanks for the reply, 
Yes the card was formatted FAT32. I also tried a 32Gb SD card.
Trying to debug this issue I have a couple of questions:

Does the hand controller need the lipo battery plugged in or should it work including TFT from the USB power alone?
What are the message about "equestFrom(): i2cRead returned Error 263"
The screen is blank is there any way to prove the screen is working? 
Can I put some debug in the code to print to the console ? If so how do I do that?

With nothing visible its really hard to see where the issue is.  What can I do next ?

Tony

**Reply by alex-code** [reply]
It'll work off USB.

Thinking if the screen and SD card isn't working the ribbon cable for the display might not be seated correctly. It can be a bit fiddly to fit.

**Reply by TonyBrock42** [reply]
Yes I had that thought and I have had a good look at it and reseated a couple of times, the other thing is the orientation. ill try to take some pics of my board
I have 3 fire beetles and 2 screens so I have spares to test but with no visual signs it so hard to figure it out.  How do I add some serial write to the code to see it on the VS Terminal ?  Is this possible and I can move this through the code maybe??

**Reply by TonyBrock42** [reply]
![EA22E080-D96D-4535-B5D6-743AA3415A25](https://github.com/DCC-EX/EX-T3-WiFi/assets/105856128/a1edd40a-74cd-4288-8407-d2e47b83c159)
![541F5A6E-0411-4D42-9495-4D858A28E86B](https://github.com/DCC-EX/EX-T3-WiFi/assets/105856128/294ebcc0-0697-4248-84c8-d6f3b969a93e)
![F32D9B65-2D16-4888-B165-9C4CF748487E](https://github.com/DCC-EX/EX-T3-WiFi/assets/105856128/72817799-179a-4ab0-988a-8677a30e59f5)
![1F488856-2FBB-427D-ACE6-8408D68D875A](https://github.com/DCC-EX/EX-T3-WiFi/assets/105856128/c1d6ca25-11b8-4959-9cf1-c64d9cec31ec)
![C80260D1-4CB2-420B-8E44-074FFAC83BAE](https://github.com/DCC-EX/EX-T3-WiFi/assets/105856128/3c13a4f8-0aed-4ccd-a7e3-6890f5702947)

**Comment by brianjackson1** [comment, marked as answer]
Looking at the picture of the ribbon connector, it looks like the ribbon is connected to the screen with the contacts facing upwards.
The contacts should be facing down on the screen connector and towards the esp32 chip on the fire beetle.

**Reply by TonyBrock42** [reply]
Ive just spotted the same, only one side of the ribbon has the connectors.  Im just swapping it,
Thanks ill let you both know.

**Comment by TonyBrock42** [comment]
Thankyou both.  That was it, never spotted the ribbon was handed.
Also the header into the ESP from encoder was 3 pins and should be 4 with a NC pin.  
Got both boards and screens working
Is full source available or are there compiled libraries?  Im looking to add a shunting screen as i have developed DCC controlled uncoupling on wagons and coaches.

Thanks again, i really appreciate your quick responses.
T

**Reply by alex-code** [reply]
It's all open source but you'll need some C++ experience to add another screen.

Does your shunting control use DCC accessory commands?

**Reply by TonyBrock42** [reply]
Hi ok great . Yes I’ve been a software developer for 30 years and C++ no problem . I’ve designed a 4 function decoder based on a pic and written the software in asm. Decoder sits under each wagon  Happy to discuss outside of GitHub

**Reply by alex-code** [reply]
Sounds like a cool project.

All the screens/displays I did for the throttle have names that end in `UI.cpp`.

Does your decoder use cab commands or accessory commands. Any ideas on the kind of control option you're wanting? Always open to adding more features.

**Reply by TonyBrock42** [reply]
Just pretends to be a loco so cab commands . I’m not 100% on the functionality but ideally the screen should show the train and which wagons / coaches it consists of and allows a coupling to be clicked to operate the coupling either side.

**Reply by alex-code** [reply]
Is the function number on your decoder re-mappable?
Something that could work now is your decoder having the same address as the loco then you could add a function button to the display.

**Reply by TonyBrock42** [reply]
Yes you can program on main only as there is no load to give the ack when on the progtamming track . Either that or I don’t understand how the data is sent to the CS when on the programming track and you request loco ID or value of a parameter

**Reply by TonyBrock42** [reply]
Just playing with the handset - looking at the web utility . So computer is connected the access point  of the handset (ssid dccex-t3-qd2m) . Webpage won’t open as dns nsme only ip address of 192.168.4.1 (pc is .2) . I can add a loco but if I edit , change tab or refresh the entry is lost , see pics 

![1720782A-E4BE-4B7B-8B77-183D17F436D1](https://github.com/DCC-EX/EX-T3-WiFi/assets/105856128/f9543e45-0cf1-4f66-8604-c66b188580ed)

then I edit 

![DE77392D-27DB-405A-A2FC-3DAEDC992DB3](https://github.com/DCC-EX/EX-T3-WiFi/assets/105856128/f85a3acb-713a-44b3-9305-ddc79cbc5b35)


![F7897B8C-C794-445F-8F88-B68AECD8C17D](https://github.com/DCC-EX/EX-T3-WiFi/assets/105856128/d55cbd21-3f94-4b54-a01a-cee8008a146a)

**Reply by alex-code** [reply]
What web browser are you using?
Could you press Ctrl+Shift+J to see if there's any errors on the console.

**Reply by TonyBrock42** [reply]
Hi I’ll check and advise . 
however I think there are 2 errors here . I have a second handset with the same code on it and that does allow the loco definitions to be saved . The issue with access to web server via nsme fails for both and the ip of x.x.4.1 is used . 

I have another issue which is present on both hand sets in that I cannot enable power to either main or prog track. The icon becomes highlighted but if you leave the main screen and return then the highlight goes out . 

If you enable power via the CS serial console with <1 main> or <1 prog> then the button highlights automatically on both hand sets . The highlight is persistent and allows control .  Control and highlight is removed with a <0 main> for example 
 Any ideas and thanks 
Tony

**Reply by alex-code** [reply]
Could you swap the SD cards to see if that could be the issue.

Do you know your CS version?
Is the throttle connected directly to the CS or do you have another interface like loconet?

**Reply by TonyBrock42** [reply]
Hi , 
Ok so I powered up the working handset (8gb Sd card) and confirmed that I could edit a loco via the webpage 192.168.4.1 when connected to the handset access points . I powered down and fitted the Sd card (32gb sandisk) from the handset that won’t save changes (just as a note that controller also failed to save the CS connection info ) . With the card installed it stll would not save. 
I took a new 4gb card and formatted to fat32 installed it into the working controller and couldn’t save wifi details or settings via website . 
I then copied m the files from the working Sd to the new 4gb and all is working ok 
I wonder why it worked and also a question would be the 2 cards now have the same ssid for the access point so this is not ideal , can the config be manually edited to make a new ssid ? 

Additionally I am on the latest CS v5.0.7 
Tony

**Reply by alex-code** [reply]
Not sure about the SD issues.
One thing that could help try and track down the issue would be having the throttle connected via USB and running the serial monitor in VS Code from PlatformIO.
That should output some debug info and possibly error info.

I'll try with a fresh formatted card and see if I have any issues.

Normally on first setup the throttle would generate a random SSID and password and save it to the `settings.json` file so copying would duplicate.
The settings file is just plain text so you can edit it with notepad but requires a little familiarity with the JSON format.
If you didn't want to edit you could rename the file and the throttle would then try to create it again.

**Reply by TonyBrock42** [reply]
Hi thanks for the reply.
So tonight i did upload and monitor.
Pressed On All, On Main and On Prog.  All 3 buttons turned solid as if they are on.
I pressed By Address then the 3 lines top right to return to the main screen all power buttons are now turned off

This was the terminal output


--- Terminal on COM4 | 115200 8-N-1
--- Available filters and text transformations: colorize, debug, default, direct, esp32_exception_decoder, hexlify, log2file, nocontrol, printable, send_on_enter, time
--- More details at https://bit.ly/pio-monitor-filters
--- Quit: Ctrl+C | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H
ets Jul 29 2019 12:21:46

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13104
load:0x40080400,len:3036
entry 0x400805e4
[   222][E][esp32-hal-cpu.c:110] addApbChangeCallback(): duplicate func=0x40104738 arg=0x3ffbde68
155768
156764
156748
155888
155884
155884
156748

As previously mentioned if i turn the power on in the CS terminal the correct button is toggled on - on the handset.

Could this be an issue with the latest release of DC EX?

Tony

**Reply by TonyBrock42** [reply]
Ok. scrap that.  I re-uploaded the CS and now works on both handsets.  Very odd

**Reply by alex-code** [reply]
There is some extra functionality which requires adding a file to the CS code and re-uploading but the throttle should still work without it.

Did you get the SD cards working too?

**Reply by TonyBrock42** [reply]
Alex can you drop me an email to maverikt@icloud.com please

**Reply by alex-code** [reply]
For support it's best to stick with GitHub or the discord server as other members may be able to help or be helped too.

https://discord.gg/y2sB4Fp
